specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+  - Update L2StandardBridge Proxy ERC-1967 Implementation Slot
+
+### L1Block Deployment
+
+- `from`: `0x4210000000000000000000000000000000000002`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2CrossDomainMessenger Deployment
+
+- `from`: `0x4210000000000000000000000000000000000003`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2StandardBridge Deployment
+
+- `from`: `0x4210000000000000000000000000000000000004`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L1Block Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
+- `to`: `0x4200000000000000000000000000000000000015`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `50,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2CrossDomainMessenger Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
+- `to`: `0x4200000000000000000000000000000000000007`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `50,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2StandardBridge Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
+- `to`: `0x4200000000000000000000000000000000000010`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `50,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+## Selection of `ETHER_TOKEN_ADDRESS`
+
+It was decided to use `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` to represent `ether` to push ecosystem standardization
+efforts around using this address to represent `ether` in [DeFi protocols](https://twitter.com/transmissions11/status/1775525755860169010)
+or [APIs](https://github.com/ethereum/execution-apis/pull/484).
+
+## Security Considerations
+
+### OptimismPortal Token Allowance
+
+The `OptimismPortal` makes calls on behalf of users. It is therefore unsafe to be able to call the address of the custom gas
+token itself from the `OptimismPortal` because it would be a simple way to `approve` an attacker's balance and steal the
+entire ERC20 token balance of the `OptimismPortal`.
+
+### Decimal Scaling
+
+A library for scaling `uint256` representations of floating point numbers is used on both ERC20 token deposits and withdrawals.
+This ensures that from the point of view of the L2 system, it always appears as if the number of tokens that were deposited
+are the same as the native balance. Tokens with less than or 18 decimals are supported, meaning that there is no potential for
+loss of precision due to sending in too small of a unit.
+
+### Interoperability Support
+
+Interop is supported between chains that use a custom gas token. The token address and the number of decimals are legible on
+chain. In the future we may add the ability to poke a chain such that it emits an event that includes the custom gas token address
+and its number of decimals to easily be able to introspect on the native asset of another chain.
```

Comment:
If we are certain we're sticking with "events as messages" for interop (compared to e.g. state), what do you think of adding `emit GasPayingTokenSet(address indexed token, uint8 indexed decimals)` now, as part of the L1Block setter? It's very simple to have the event always emitted once when set, and no one ever needs to poke

Reply:
One consideration is that we will likely need to have an "event horizon" where events older than a certain age cannot be used as an initiating message. This is because of memory restrictions in the fault proof. The fault proof needs to reproduce the event via execution, and it would need to walk back a ton of blocks for really old events. Which is why the poke pattern would be required. Not opposed to adding the event now though for clarity, but feels premature to introduce the poking method now

Reply:
Interesting! Did not realize that event horizon requirement. 

> it would need to walk back a ton of blocks for really old events

What does this mean exactly and how does that result in large memory execution compared to more recent logs?

Either way, given that, I agree my suggestion is premature and poking may be preferable. Ideally we have just one way to emit the event for simplicity/fewer code paths. So let's just remember to track this open question somewhere.

Reply:
The fault proof program walks backwards from a L2 blockhash that is on L1 until its able to have all of the inputs (L2 block data) to re-execute to recreate the data required to ensure that the proposed commit is "correct". If a log from a really long time ago is used as part of an executing message, then we will need to walk back to the point where that log is created as a side effect of evm execution

Reply:
Will follow up with adding an event that is emitted during the call to set and leave out the ability to poke and have the event emitted again. See https://github.com/ethereum-optimism/protocol-quest/issues/222#issuecomment-2041782114

Reply:
Addressed in https://github.com/ethereum-optimism/optimism/pull/10011/commits/d7d8ad71e192b7e995176776d4664a2e56f54e80